lib/generate.js

```diff
@@ -104,47 +107,62 @@ function generateLayoutInternal(options, callback) {
         });
 
         /* Only keep 1 item per svg signature for packing */
-        options.imgs = options.imgs.filter(function(item) {
+        options.imgs = options.imgs.filter((item) => {
             var svg = svgPerItemId[item.id];
             return item.id === itemIdsPerSvg[svg][0];
         });
     }
 
-    function createImagesWithSize(img, callback) {
+    function encodeImage(image, callback) {
+        image.encode('png', (err, buffer) => {
+            if (err) return callback(err);
+            callback(null, {
+                width: image.width(),
+                height: image.height(),
+                buffer
+            });
+        });
+    }
+
+    function createImages(img, callback) {
         var mapnikOpts = { scale: options.pixelRatio };
         if (options.maxIconSize) {
             mapnikOpts.max_size = options.maxIconSize;
         }
-        mapnik.Image.fromSVGBytes(img.svg, mapnikOpts, function(err, image) {
+        mapnik.Image.fromSVGBytes(img.svg, mapnikOpts, (err, image) => {
             if (err && err.message.match(/image created from svg must be \d+ pixels or fewer on each side/) && options.removeOversizedIcons) return callback(null, null);
             // Produce a null result if no width or height attributes. The error message from mapnik has a typo "then"; account for potential future fix to "than".
             if (err && err.message.match(/image created from svg must have a width and height greater (then|than) zero/)) return callback(null, null);
             if (err) return callback(err);
             if (!image.width() || !image.height()) return callback(null, null);
-            image.encode('png', function(err, buffer) {
+
+            let q = new queue();
+            q.defer(encodeImage, image);
+            if (options.stretchMetadata && img.svg.includes('mapbox-stretch')) {
```

Comment:
Could slightly improve performance by also checking for `options.format`.

When `options.format` is not `true`, we generate an actual image.
When it is `true`, we generate a layout JSON. Only the layout JSON needs to have the stretch metadata.

Reply:
Good call! I'll make that change.